### ISSUE TITLE
NextJS ES Module build fails because of react-dnd

### ISSUE THREAD

**Issue by ptakpatryk** [react-querybuilder/react-querybuilder#340]
It all works fine with the barebone React app although fails with the NEXT.JS app.
Works with version 4.2.3 where it uses an older version of react-dnd:
`react-dnd "^14.0.4"
    react-dnd-html5-backend "^14.0.2"`

Please take a look at this sandbox.
https://codesandbox.io/s/next-querybuilder-pj1d4s

Error msg:
![image](https://user-images.githubusercontent.com/42934159/176429153-1413daf1-4371-4ff0-a1b1-0d380f89ad36.png)

**Comment by jakeboone02** [comment]
React-DnD is ES-only now, but it's still weird that Next considers RQB's `dist/index.cjs.js` to be a module since that build specifies "cjs". What happens if you use `<QueryBuilderWithoutDndProvider />` wrapped with your own `<DndProvider />`?

**Comment by ptakpatryk** [comment]
Thanks for the response @jakeboone02 
Sadly exactly the same. Yeah I found it weird that it reaches for the common js module by default.

**Comment by jakeboone02** [comment]
After a quick review, it appears like the CommonJS build of RQB is never going to work as currently configured since it tries to `require()` an ESM-only version of React-DnD. I think we only have two options:

1. Go ESM-only with RQB. I've thought about doing this before, but never felt like it was worth a major version bump since it wasn't really any extra trouble to maintain a CommonJS build configuration. (Also, under the assumption that RQB is used mostly in corporate/enterprise environments, we try to support the legacy stuff like IE11 rather than force users to update their environments over which they may have little or no control.)
2. Revert the React-DnD dependency to the latest version that supports CommonJS (i.e. `react-dnd@^15`).

Since you're the first to bring this up, and RQB has had React-DnD v16 as a dependency since v4.3.1 (released over a month ago), I'm going to assume that not many people are using it in a CommonJS environment. That would make option 1 more attractive, however, the difference between React-DnD v15 and v16 is really just that v16 is ESM-only. That makes option 2 more attractive, for now anyway.

I'll need to see what happens when an app has a direct dependency on `react-dnd@^16` and a transitive dependency (via RQB) on `react-dnd@^15`. Stay tuned.

**Comment by mohsinamjad** [comment]
I'm also facing the same issue sadly,

**Comment by jakeboone02** [comment]
@ptakpatryk / @mohsinamjad do either of you have a direct dependency on `react-dnd`? If so, which version are you using?

I ask because it's looking like I need to revert RQB's `react-dnd` dependency to `@^15`. That will break any RQB implementations that use RQB's `<QueryBuilderWithoutDndProvider />` and separately import `react-dnd@16`.

**Comment by ptakpatryk** [comment]
@jakeboone02 I did end up using react-querybuilder@4.2.3 and didn't install react-dnd at all even though I'm using it. I import `react-dnd` without actually installing it in my root folder so I presume it uses the one that was installed as a dependency of RQB. I tried with a different react-dnd version and couldn't make it work by using `<QueryBuilderWithoutDndProvider />` and my separate provider. I was getting an error that it needs a provider.

What would be good (not sure if possible) to strip off react-dnd completely if `enableDragAndDrop` is set to be `false`. I'm not using it at all but having problems because of it ^^
Thanks for looking into it Jake, really appreciate :)

**Comment by jakeboone02** [comment]
> I import `react-dnd` without actually installing it in my root folder so I presume it uses the one that was installed as a dependency of RQB.

Yep, that method has its own issues and is generally not recommended, but it can actually work well to keep your transitive dependency versions compatible with your direct dependencies.

> What would be good (not sure if possible) to strip off `react-dnd` completely if `enableDragAndDrop` is set to be `false`.

Yeah, that would certainly be best. I'm not currently sure how to do it considering the way `react-dnd` hooks and refs work, but it's definitely worth looking into.

Thanks for the feedback!

**Comment by mohsinamjad** [comment]
@jakeboone02 I don't have any direct react-dnd dependency, I also installed version 4.2.3 which works fine I guess.

**Comment by jakeboone02** [comment]
Thanks, @mohsinamjad.

I might try moving `react-dnd` (and `react-dnd-html5-backend`) to `devDependencies` and `peerDependencies`, using `peerDependenciesMeta` to make them optional for consuming applications. Then we'd just have to figure out how to detect them and fall back to `enableDragAndDrop={false}` (with a dev-only console warning) when they're not available. This also probably means we'll have to bite the bullet and go ESM-only.

**Comment by jakeboone02** [comment]
@ptakpatryk / @mohsinamjad - I created PR #343 which attempts to make `react-dnd` an optional dependency for `react-querybuilder`. It doesn't currently work unless the consuming application has `react-dnd` installed, but it will inherit the version of `react-dnd` that the consumer _does_ have installed. So...baby steps.

[This codesandbox](https://codesandbox.io/s/next-querybuilder-forked-465ps5?file=/pages/index.js) demonstrates `react-querybuilder` as of #343 with `react@17` and `react-dnd@15`.

**Comment by Adnan0061** [comment]
@jakeboone02 
hi, it appears that it's the same for me also though I am not even using react-dnd

I am using 
next- 12.2.2
react- 18.2.0
dnd-kit/core- 6.0.5
dnd-kit/sortable- 7.0.1

![image](https://user-images.githubusercontent.com/48943431/184119943-11313157-f159-415e-9cd7-e9eaa9f10b63.png)

**Comment by jakeboone02** [comment]
@Adnan0061 yeah, the problem is not with ReactDnD itself, but with the CommonJS build of RQB (see https://github.com/react-querybuilder/react-querybuilder/issues/340#issuecomment-1170381138).

**Comment by Adnan0061** [comment]
I saw, also checked the PR. but why isn't it merging? it's been a month

> @Adnan0061 yeah, the problem is not with ReactDnD itself, but with the CommonJS build of RQB (see [#340 (comment)](https://github.com/react-querybuilder/react-querybuilder/issues/340#issuecomment-1170381138)).

**Comment by jakeboone02** [comment]
> why isn't it merging?

Because I can't figure out how to get it to work without requiring every dependent application to install `react-dnd` (and `react-dnd-html5-backend`) whether or not they actually use the drag-and-drop feature.

I want `react-dnd` to be a _truly_ optional dependency. Install it if you need it, otherwise you don't even need to think about it. If that's an impossible goal, I want someone to prove that to me before I abandon the effort.

Maybe I could do it with a plugin system of come kind, or a wrapper package (`@react-querybuilder/dnd`?). Hmmm...

> it's been a month

Timelines work a little differently here in the very-small-open-source-library world. I sincerely hope this doesn't come across as complaining or lashing out or anything like that, and I'm sure you were just innocently curious, _but_... I'm effectively the sole maintainer of `react-querybuilder`; this is a side project; I have a wife and kids I like spending time with; it's summer here in the northern hemisphere so we've been on vacation recently in places where I didn't even want to think about computers; ...etc. **TL;DR: life.**

On top of all that, I've received no real help actually _solving_ the issue. Even my [StackOverflow question](https://stackoverflow.com/questions/72862593/how-to-publish-a-js-package-that-gracefully-fails-if-an-optional-peer-dependency) only has 59 views and one attempted answer. If someone could point me in the right direction I'd love to head that way. It _does_ bother me that RQB doesn't work with such a popular tool as NextJS, and I genuinely want to fix it ASAP! (Though the 'P' in 'ASAP' may stand for "practical" and not "possible".)

**Comment by Adnan0061** [comment]
don't get me wrong, It's just an expression because we couldn't use this amazing package because of this. 
personally, I think this is still amazing without the drag-and-drop option. I support the plugin system for dnd so that those who need it will use it, might be a good option rather than keeping it unusable. 

If the time quote hurt your feelings, I am truly sorry for that.

> > it's been a month
> 
> Timelines work a little differently here in the very-small-open-source-library world. I sincerely hope this doesn't come across as complaining or lashing out or anything like that, and I'm sure you were just innocently curious, _but_... I'm effectively the sole maintainer of `react-querybuilder`; this is a side project; I have a wife and kids I like spending time with; it's summer here in the northern hemisphere so we've been on vacation recently in places where I didn't even want to think about computers; ...etc. **TL;DR: life.**
> 
> On top of all that, I've received no real help actually _solving_ the issue. Even my [StackOverflow question](https://stackoverflow.com/questions/72862593/how-to-publish-a-js-package-that-gracefully-fails-if-an-optional-peer-dependency) only has 59 views and one attempted answer. If someone could point me in the right direction I'd love to head that way. It _does_ bother me that RQB doesn't work with such a popular tool as NextJS, and I genuinely want to fix it ASAP! (Though the 'P' in 'ASAP' may stand for "practical" and not "possible".)

**Comment by jakeboone02** [comment]
No hurt feelings! And no need to apologize. I just wanted to be open about the reality of the situation. I appreciate your kind words about RQB.

**Comment by jakeboone02** [comment]
> Maybe I could do it with a plugin system of come kind, or a wrapper package (`@react-querybuilder/dnd`?). Hmmm...

Quick update: the wrapper package is looking promising. Nothing to show just yet, but I'll update #343 within the next few days if I can.

**Comment by jakeboone02** [comment]
Ok I think I finally have #343 done. It extracts all drag-and-drop code into a separate package called `@react-querybuilder/dnd`, so using `react-querybuilder` v5 by itself should avoid any `react-dnd`-related problems (as it pertains to this library). It also makes the `react-dnd` dependency dynamic so you can use any version >=14.

Currently the new package exports a drop-in replacement for `<QueryBuilder />` that includes drag-and-drop functionality. But I realized last night that it could have used composition instead.

**I need to make a choice between the following two upgrade paths from v4 to v5.**

(**Note:** These changes will _only_ apply if `enableDragAndDrop` is `true` - otherwise the upgrade won't require any code changes. When applicable, both options will require users to manually add `react-dnd` and `react-dnd-html5-backend` since they will be `peerDependencies` of `@react-querybuilder/dnd`.)

### Option 1 (current design)

```diff
-import { QueryBuilder } from 'react-querybuilder';
+import { QueryBuilder } from '@react-querybuilder/dnd';

 export const App = () => {
   return (
     <QueryBuilder enableDragAndDrop />
   )
 }
```

### Option 2

```diff
+import { QueryBuilderDnD } from '@react-querybuilder/dnd';
 import { QueryBuilder } from 'react-querybuilder';

 export const App = () => {
   return (
-    <QueryBuilder enableDragAndDrop />
+    <QueryBuilderDnD>
+      <QueryBuilder enableDragAndDrop />
+    </QueryBuilderDnD>
   )
 }
```

Option 2 would simplify `@react-querybuilder/dnd`'s code a bit, but it's _slightly_ more work for the users.

Any opinions on which way is better? I could design it to allow either method, but that wouldn't help with simplifying this library's code like if I chose option 2 by itself. (The code simplification isn't super important since the more complex version is already written, after all.)

**Comment by MattEagle95** [comment]
Hey Jake! Thanks for the work on this one, especially in summer 🌴. I'm exited to use it in NextJS!

I just wanted to give my opinion. I like option 2, because it extends the current `QueryBuilder` instead of replacing it with another instance of another package. This adds a bit more "boilerplate" code to it, but I think when it's written like that, it's easier to understand. I dont use "another" `QueryBuilder`, I use the same one, enable it's internal drag-and-drop functionality by setting the `enableDragAndDrop` prop and wrap it in a DnD context. Maybe your just written DnD-context, maybe a context I customized myself, its extendable.

I think this will allow for easier customization and extension in the future. Imagine another `wrapper` package for a new dnd-library or a `QueryBuilderLocaleProvider`/`QueryBuilderDateFns` something like that. With option 2 I could just add or remove wrappers around the `QueryBuilder`. With option 1 I can only use 1 library, because im using that one's whole `QueryBuilder`.

Option 1
```
import { QueryBuilder } from '@react-querybuilder/dnd';

<QueryBuilder enableDragAndDrop enableLocaleProvider /> // enableLocaleProvider functionality needs to be in dnd package, because I use it's QueryBuilder
```

Option 2
```
import { QueryBuilderLocaleProvider} from '@react-querybuilder/cool-locale-stuff';
import { QueryBuilderDnD } from '@react-querybuilder/dnd';
import { QueryBuilder } from 'react-querybuilder';

<QueryBuilderLocaleProvider>
    <QueryBuilderDnD>
        <QueryBuilder enableDragAndDrop />
    </QueryBuilderDnD>
</QueryBuilderLocaleProvider>
```

I hope you can understand what I mean : ) I'm no professional in writing libraries, just wanted to share.
Whatever it will be, both options are alright. Have a nice day! : )

**Comment by jakeboone02** [comment]
Gah! That makes so much sense!

Thanks so much for the excellent feedback. I wasn't even considering the possibility of *other* enhancement/augmentation/wrapper packages, but now that seems like the obvious path forward. (Can't wait to see what `@react-querybuilder/cool-locale-stuff` does! ;) )

I'll refactor the PR to support composition and start prepping for v5!

**Comment by jakeboone02** [comment]
All - I just merged #343 after confirming that it works with NextJS where v4.5.2 fails. Thanks for all your help!

I'm going to release v5.0 soon (maybe as a beta first to be careful), but I have one more major decision to make: should we go [ESM-only](https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c)? I'm not aware of any problems the current build setup is causing (outside of this issue, which is resolved of course), but it seems right to embrace The Future™ anyway. If someone needs a CommonJS or UMD version they can always stay on v4, which isn't really that different from v5 in terms of functionality at this point. Anyone have any opinions on it?
